examples/sitemap/app/lib/sitemap.ts

```diff
@@ -0,0 +1,335 @@
+import {LoaderFunctionArgs} from '@shopify/remix-oxygen';
+import {
+  CountryCode,
+  LanguageCode,
+} from '@shopify/hydrogen/storefront-api-types';
+
+type Locale = `${LanguageCode}-${CountryCode}`;
+
+type SITEMAP_INDEX_TYPE =
+  | 'pages'
+  | 'products'
+  | 'collections'
+  | 'blogs'
+  | 'articles'
+  | 'metaObjects';
+
+/**
+ * Generate a sitemap index that links to separate sitemaps for each resource type.
+ */
+export async function getSitemapIndex({
+  storefront,
+  request,
+  types = [
+    'pages',
+    'products',
+    'collections',
+    'metaObjects',
+    'pages',
+    'articles',
+  ],
+  customUrls = [],
+}: {
+  storefront: LoaderFunctionArgs['context']['storefront'];
+  request: Request;
+  types?: SITEMAP_INDEX_TYPE[];
+  customUrls?: string[];
+}) {
+  const data = await storefront.query(SITEMAP_INDEX_QUERY, {
+    storefrontApiVersion: 'unstable',
+  });
+
+  if (!data) {
+    throw new Response('No data found', {status: 404});
+  }
+
+  const baseUrl = new URL(request.url).origin;
+
+  const body = `<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">
+${types
+  .map((type) => getSiteMapLinks(type, data[type].pagesCount.count, baseUrl))
+  .join('\n')}
+${customUrls
+  .map(
+    (url) => `<sitemap>
+  <loc>${url}</loc>
+</sitemap>`,
+  )
+  .join('\n')}
+</urlset>`;
+
+  return new Response(body, {
+    headers: {
+      'Content-Type': 'application/xml',
+      'Cache-Control': `max-age=${60 * 60 * 24}`,
+    },
+  });
+}
+
+type GetSiteMapOptions = {
+  /** The params object from Remix */
```

Comment:
I think types with comment annotation like this don't work well in our transpiling logic for JS projects. `interface` works well though, so perhaps better to use that in this case.

I want to dedicate time to improve the transpiler to JSDoc and cover this case.. some day 😓